### PR TITLE
migrate to new SELinux api

### PR DESCRIPTION
**Pull request by cgzones**
Using hard-coded access vector ids is deprecated and can lead to issues with custom SELinux policies.
Switch to `selinux_check_access()`.

Also use the libselinux log callback and log if available to audit.
This makes it easier for users to catch SELinux denials.

Drop legacy shortcut logic for passwd, which avoided a SELinux check if uid 0 changes a password of a user which username equals the current SELinux user identifier.
Nowadays usernames rarely match SELinux user identifiers and the benefit of skipping a SELinux check is negligible.

Closes: #182

### CODE REVIEW

File: src/passwd.c [hunk 1]
```diff
@@ -740,17 +736,7 @@ static int check_selinux_access (const char *changed_user,
 	if (changed_uid != 0 && strcmp(changed_user, user) == 0) {
```
Comment (cgzones) [inline src/passwd.c]: `if (changed_uid != 0 && strcmp(changed_user, user) == 0)`
Is this shortcut really necessary?
Also the comment above does not match.

**Comment by cgzones** [comment]
I asked on the [SELinux mailing](https://lore.kernel.org/selinux/CAJ2a_Dd=T1sQh5zCqJLf8XTSQVU6Qgz78aFpT2oQ30VYVbQnWw@mail.gmail.com/) list for feedback

File: lib/selinux.c [hunk 2]
```diff
@@ -98,6 +98,50 @@ int reset_selinux_file_context (void)
...
+
+	/* caller has checked for (is_selinux_enabled () > 0) already */
+
+	passwd_class_id = string_to_security_class ("passwd");
+	if (0 == passwd_class_id) {
+		fprintf (stderr,
+			    _("%s: can not find SELinux class 'passwd'\n"),
+			    Prog);
+		SYSLOG ((LOG_WARN, "can not find SELinux class 'passwd'"));
+		if (0 != security_getenforce ()) {
```
Comment (stephensmalley) [inline lib/selinux.c]: Checking security_deny_unknown() would be more appropriate here, and is what libselinux selinux_check_access() does.

File: lib/selinux.c [hunk 3]
```diff
@@ -98,6 +98,50 @@ int reset_selinux_file_context (void)
...
+		}
+	}
+
+	permission_id = string_to_av_perm (passwd_class_id, perm_name);
+	if (0 == permission_id) {
+		fprintf (stderr,
+			    _("%s: can not find '%s' SELinux access vector for class 'passwd'\n"),
+			    Prog, perm_name);
+		SYSLOG ((LOG_WARN, "can not find '%s' SELinux access vector for class 'passwd'", perm_name));
+		if (0 != security_getenforce ()) {
```
Comment (stephensmalley) [inline lib/selinux.c]: Ditto

File: lib/selinux.c [hunk 4]
```diff
@@ -98,6 +98,50 @@ int reset_selinux_file_context (void)
...
+			closelog ();
+			exit (E_SELINUX_ERR);
+		} else {
+			return 0;
+		}
+	}
+
+	return selinux_check_passwd_access (permission_id);
```
Comment (stephensmalley) [inline lib/selinux.c]: Wondering if we ought to just convert to direct use of libselinux selinux_check_access().  That would require calling getprevcon() here and passing it as the source and target contexts, but avoids mapping the class twice.  If so, would need/want to also use selinux_set_callback() to set the logging callback and ideally wire that up to audit.

File: src/passwd.c (hunk 1 above)
Comment (stephensmalley) [reply src/passwd.c]: This logic appears to be a legacy of original SELinux reference implementation where SELinux user identity (i.e. first field of the security context) was the same as the Linux username.  Then the logic tries to require that in order to change the password of a different user, one needs to have the "passwd" permission.  Then someone further amended it to also require the "passwd" permission if changing the root password regardless. Seems like this will almost always fail since the SELinux user identity is rarely (never?) the same as the Linux username and thus we will always execute the else clause.  Likely should be removed.

**Comment by stephensmalley** [comment]
For new code, try to use selinux_check_access() and let libselinux internally handle the string_to_security_class, string_to_av_perm, security_deny_unknown, security_getenforce, etc calls and handling.  See dbus-broker for an example.

**Comment by cgzones** [comment]
updated to use `selinux_check_access()` and enabled logging via `selinux_set_callback(SELINUX_CB_LOG, ...)` (also if available inform audit)

File: lib/selinux.c (hunk 2 above)
Comment (cgzones) [reply lib/selinux.c]: switched to `selinux_check_access()`

File: lib/selinux.c (hunk 3 above)
Comment (cgzones) [reply lib/selinux.c]: switched to `selinux_check_access()`

File: lib/selinux.c [hunk 5]
```diff
@@ -98,6 +98,136 @@ int reset_selinux_file_context (void)
...
+	va_end (ap);
+
+	if (r < 0) {
+		free (buf);
+		return 0;
+	}
+
+	fprintf (stderr, _("%s: libselinux: %s\n"),
+			    Prog, buf);
+	SYSLOG ((LOG_WARN, "libselinux: %s", buf));
```
Comment (stephensmalley) [inline lib/selinux.c]: Do we want to always do both (log + audit) or only fall back to SYSLOG if audit not enabled?

File: lib/selinux.c [hunk 6]
```diff
@@ -98,6 +98,136 @@ int reset_selinux_file_context (void)
...
+		    Prog, strerror (errno));
+		SYSLOG ((LOG_WARN,
+		    "can not get previous SELinux process context: %s",
+		    strerror (errno)));
+		freecon (user_context_str);
+		return (security_getenforce () != 0);
+	}
+
+	if (changed_user) {
+		/* get the "user" portion of the context
```
Comment (stephensmalley) [inline lib/selinux.c]: I am of the opinion that this entire if statement and its associated dependencies can just go away, i.e. no shortcut logic because SELinux user identity will generally never be the same as Linux username on any modern Linux distro. Worst case is that we'll trigger a passwd permission check that wasn't truly needed.

File: src/chfn.c [hunk 7]
```diff
@@ -280,7 +276,7 @@ static void process_flags (int argc, char **argv)
 		{NULL, 0, NULL, '\0'}
 	};
 
-	/* 
+	/*
 	 * The remaining arguments will be processed one by one and executed
```
Comment (stephensmalley) [inline src/chfn.c]: Usually you don't make unrelated whitespace fixes in the same commit as a real change, i.e. if you do them at all, do them in a separate commit.

File: src/passwd.c [hunk 8]
```diff
@@ -331,7 +320,7 @@ static int new_password (const struct passwd *pw)
 
 		/*
 		 * If enabled, warn about weak passwords even if you are
-		 * root (enter this password again to use it anyway). 
+		 * root (enter this password again to use it anyway).
```
Comment (stephensmalley) [inline src/passwd.c]: ditto

File: src/passwd.c [hunk 9]
```diff
@@ -1034,22 +977,15 @@ int main (int argc, char **argv)
...
-		if (getprevcon (&user_context) == 0) {
-			user = user_context; /* FIXME: use context_user_get? */
-		}
+	current_uid = getuid() ;
+	if ((0 == current_uid) &&
+	    (check_selinux_permit_shortcut ("passwd", name, pw->pw_uid) != 0)) {
 		SYSLOG ((LOG_ALERT,
-		         "%s is not authorized to change the password of %s",
-		         user, name));
+		         "uid %d is not authorized by SELinuxto change the password of %s",
```
Comment (stephensmalley) [inline src/passwd.c]: Missing a space.

File: lib/selinux.c [hunk 10]
```diff
@@ -98,6 +98,136 @@ int reset_selinux_file_context (void)
...
+	r = vasprintf (&buf, fmt, ap);
+	va_end (ap);
+
+	if (r < 0) {
+		free (buf);
+		return 0;
+	}
+
+	fprintf (stderr, _("%s: libselinux: %s\n"),
+			    Prog, buf);
```
[the remark this answers is not in the document]
Comment (cgzones) [reply lib/selinux.c]: dropped the `stderr` logging

File: lib/selinux.c (hunk 5 above)
Comment (cgzones) [reply lib/selinux.c]: switched to try `audit`, use `syslog` only as fallback

File: lib/selinux.c [hunk 11]
```diff
@@ -98,6 +98,136 @@ int reset_selinux_file_context (void)
 	return 0;
 }
 
+/*
+ *  libselinux uses this callback when access gets denied or other
+ *  events happen. If audit is turned on, messages will be reported
+ *  using audit netlink, otherwise they will be logged using the usual
+ *  channels.
+ *
+ *  Code adopted from systemd, which adopted from dbus.
```
[the remark this answers is not in the document]
Comment (cgzones) [reply lib/selinux.c]: The code should be different enough to not cause problems, but I am not a copyright expert.
The systemd code can be found [here](https://github.com/systemd/systemd/blob/c8966bffdd9ad4ef7dcec89e0cc8d43529ef2aed/src/core/selinux-access.c#L96)

File: src/chfn.c (hunk 7 above)
Comment (cgzones) [reply src/chfn.c]: dropped

File: src/chfn.c [hunk 12]
```diff
@@ -391,7 +386,7 @@ static void check_perms (const struct passwd *pw)
 	/*
 	 * Non-privileged users are optionally authenticated (must enter the
 	 * password of the user whose information is being changed) before
-	 * any changes can be made. Idea from util-linux chfn/chsh. 
+	 * any changes can be made. Idea from util-linux chfn/chsh.
```
[the remark this answers is not in the document]
Comment (cgzones) [reply src/chfn.c]: dropped

File: src/passwd.c (hunk 8 above)
Comment (cgzones) [reply src/passwd.c]: dropped

File: src/passwd.c (hunk 9 above)
Comment (cgzones) [reply src/passwd.c]: fixed

File: lib/selinux.c [hunk 13]
```diff
@@ -98,6 +98,136 @@ int reset_selinux_file_context (void)
...
+                                      uid_t changed_uid)
+{
+	char *user_context_str;
+	int r;
+
+	if (0 == is_selinux_enabled ()) {
+		return 0;
+	}
+
+	selinux_set_callback (SELINUX_CB_LOG, (union selinux_callback) log_callback);
```
[the remark this answers is not in the document]
Comment (cgzones) [reply lib/selinux.c]: The running applications (passwd, chfn,...) should only check for SELinux permit once in their runtime, so the question is, is the overhead worth avoiding additionally pointer assignments?

File: lib/selinux.c [hunk 14]
```diff
@@ -98,6 +99,146 @@ int reset_selinux_file_context (void)
...
+#ifdef WITH_AUDIT
+	static int selinux_audit_fd = -2;
+#endif
+
+	va_start (ap, fmt);
+	r = vasprintf (&buf, fmt, ap);
+	va_end (ap);
+
+	if (r < 0) {
+		free (buf);
```
Comment (stephensmalley) [inline lib/selinux.c]: buf could be uninitialized here?

File: lib/selinux.c [hunk 15]
```diff
@@ -98,6 +99,146 @@ int reset_selinux_file_context (void)
...
+		if (   (errno != EINVAL)
+		    && (errno != EPROTONOSUPPORT)
+		    && (errno != EAFNOSUPPORT)) {
+
+		    (void) fputs (_("Cannot open audit interface.\n"),
+		              stderr);
+		    SYSLOG ((LOG_WARN, "Cannot open audit interface."));
+		}
+	} else {
+		uid_t audit_uid = geteuid ();
```
Comment (stephensmalley) [inline lib/selinux.c]: Is this the right uid to use here?  Should we be using the real UID, or the login UID, or just pass 0 like systemd?

File: lib/selinux.c [hunk 16]
```diff
@@ -98,6 +99,146 @@ int reset_selinux_file_context (void)
...
+	selinux_set_callback (SELINUX_CB_LOG, (union selinux_callback) selinux_log_cb);
+
+	if (getprevcon (&user_context_str) != 0) {
+		fprintf (stderr,
+		    _("%s: can not get previous SELinux process context: %s\n"),
+		    Prog, strerror (errno));
+		SYSLOG ((LOG_WARN,
+		    "can not get previous SELinux process context: %s",
+		    strerror (errno)));
+		freecon (user_context_str);
```
Comment (stephensmalley) [inline lib/selinux.c]: No need to free here; getprevcon() failed.  And user_context_str could be uninitialized.

File: lib/selinux.c (hunk 6 above)
Comment (stephensmalley) [reply lib/selinux.c]: Do you think we need this logic?

File: src/passwd.c [hunk 17]
```diff
@@ -1034,22 +977,15 @@ int main (int argc, char **argv)
...
 	   changing a password without entering the old one */
-	if ((is_selinux_enabled() > 0) && (getuid() == 0) &&
-	    (check_selinux_access (name, pw->pw_uid, PASSWD__PASSWD) != 0)) {
-		security_context_t user_context = NULL;
-		const char *user = "Unknown user context";
-		if (getprevcon (&user_context) == 0) {
-			user = user_context; /* FIXME: use context_user_get? */
-		}
+	current_uid = getuid() ;
+	if ((0 == current_uid) &&
```
Comment (stephensmalley) [inline src/passwd.c]: We should probably be checking the "amroot" variable used elsewhere for consistency.  Otherwise, if the amroot logic changes later, we won't match.

File: lib/selinux.c (hunk 14 above)
Comment (cgzones) [reply lib/selinux.c]: fixed, thanks

File: lib/selinux.c (hunk 16 above)
Comment (cgzones) [reply lib/selinux.c]: fixed, thanks

File: lib/selinux.c (hunk 6 above)
Comment (cgzones) [reply lib/selinux.c]: I agree with removing this logic; I'd like to get an OK from upstream.
Should I announce this change on the mailing-list?

File: lib/selinux.c (hunk 15 above)
Comment (cgzones) [reply lib/selinux.c]: Good question; From an SELinux standpoint the uid probably does not matter at all (only the context).

File: src/passwd.c (hunk 17 above)
Comment (cgzones) [reply src/passwd.c]: The result of `0 == current_uid` (where `current_uid = getuid()`) should be the same as the value `amroot`. So simplify the whole condition to (ignoring the shortcut logic for now)?
```
if (amroot && (check_selinux_permit_shortcut ("passwd", name, pw->pw_uid) != 0)) {
```

File: lib/selinux.c (hunk 15 above)
Comment (stephensmalley) [reply lib/selinux.c]: Looks like dbus-broker uses getuid() i.e. real UID.  For shadow utils, that would be the UID of the caller even if the util is setuid-root, so that seems preferable to using geteuid().  Or you could ask what uid to use on linux-audit mailing list.

File: lib/selinux.c (hunk 6 above)
Comment (stephensmalley) [reply lib/selinux.c]: Sure, you could ask there. May also want to compare with source code of passwd, which is a separate package in Fedora?

File: src/passwd.c (hunk 17 above)
Comment (stephensmalley) [reply src/passwd.c]: Correct.  That way if someone later alters the amroot logic we'll pick up the change automatically, and keep in sync.  We want to do this check iff it was possible to change the password without entering the old one.

File: lib/selinux.c (hunk 14 above)
Comment (stephensmalley) [reply lib/selinux.c]: Still not sure it is safe/necessary to free(buf) here.  man page for vasprintf() says "If memory allocation wasn't  possible,  or  some other error occurs, these functions will return -1, and the contents of strp are undefined."  That means buf could be undefined after the call if it returns -1.

File: src/passwd.c (hunk 17 above)
Comment (cgzones) [reply src/passwd.c]: changed

File: lib/selinux.c (hunk 14 above)
Comment (cgzones) [reply lib/selinux.c]: you're absolutely right, fixed; thanks

File: src/passwd.c (hunk 1 above)
Comment (cgzones) [reply src/passwd.c]: Legacy shortcut dropped

File: lib/selinux.c (hunk 6 above)
Comment (cgzones) [reply lib/selinux.c]: Announced
(p.s.: fedora's passwd does not use something like this)

File: lib/selinux.c (hunk 15 above)
Comment (cgzones) [reply lib/selinux.c]: switched to hardcoded `0`, like systemd, dbus and fedora's passwd

**Comment by hallyn** [comment]
Thanks everyone!
